arch-arm: Rewrite the ArmTLB storage to use an AssociativeCache

giactra opened the pull request:
With this PR we replace the TlbEntry storage within the TLB from an
array of entries with a custom hardcoded FA indexing policy and LRU
replacement policy, into the flexible SetAssociative cache.

giactra committed:
arch-arm: Use the AssociativeCache in the ArmTLB

With this commit we replace the TlbEntry storage within the TLB from an
array of entries with a custom hardcoded FA indexing policy and LRU
replacement, into the flexible SetAssociative cache.

Change-Id: Ia74ff6962ac8195802b51dcc0caa516965f0ce37
Signed-off-by: Giacomo Travaglini <giacomo.travaglini@arm.com>
Reviewed-by: Andreas Sandberg <andreas.sandberg@arm.com>

giactra committed:
arch-arm: Make TlbEntry a ReplaceableEntry

Change-Id: I3b8169bfb620ea36f6bbe63c38b71184285b55c2
Signed-off-by: Giacomo Travaglini <giacomo.travaglini@arm.com>
Reviewed-by: Andreas Sandberg <andreas.sandberg@arm.com>

giactra committed:
arch-arm: Add pagesize field to the Lookup data structure

Change-Id: Ibc2c80cbf3cfd98f24440e8e6ddf4dbb7e4e26d6
Signed-off-by: Giacomo Travaglini <giacomo.travaglini@arm.com>
Reviewed-by: Andreas Sandberg <andreas.sandberg@arm.com>

giactra committed:
arch-arm: Add invalidate method to the TlbEntry

This is needed for compliance with the AssociativeCache
container. It will call the invalidate method when
invalidating the TLB entry

Change-Id: Idb1bc40b5aea8c475146700c81ab79d9980f745d
Signed-off-by: Giacomo Travaglini <giacomo.travaglini@arm.com>
Reviewed-by: Andreas Sandberg <andreas.sandberg@arm.com>

giactra committed:
arch-arm: Generate Lookup from TlbEntry

Change-Id: I355d190acfeb3cd829647b962548c82dd0013f8d
Signed-off-by: Giacomo Travaglini <giacomo.travaglini@arm.com>
Reviewed-by: Andreas Sandberg <andreas.sandberg@arm.com>

giactra committed:
arch-arm: Add insert method to the TlbEntry

Change-Id: I664b03b61e4540025c6cebaa4a7298297565c76b
Signed-off-by: Giacomo Travaglini <giacomo.travaglini@arm.com>
Reviewed-by: Andreas Sandberg <andreas.sandberg@arm.com>

giactra committed:
arch-arm: Implement the L2 TLB as a 5-way set associative

Change-Id: I65d7a384f6d54989cec3c431090c35285011849f
Signed-off-by: Giacomo Travaglini <giacomo.travaglini@arm.com>
Reviewed-by: Andreas Sandberg <andreas.sandberg@arm.com>

giactra committed:
arch-arm: Keep track of observed page sizes in the TLB

With this commit we record the page sizes of all valid
TLB entries in a TLB.

We update the set conservatively, therefore allowing
false positives but not false negatives.

This information will be used when doing a page size based
lookup. At the moment we don't strictly need it as we
iterate over all TLB entries (the TLB implements a fully
associative cache) and if we find multiple matches, it means
we have stored some partial translations.

The existing logic is prioritizing complete translations
over partial translations and among the latter, late stage
translations over early stage (with the idea to minimize
the number of walks).

The "iterate once over the entire TLB and record all matches"
won't work well when we shift from a fully associative
TLB into a set associative. With the introduction of the
aforementioned set, we can do page size based lookups,
so we can explicitly lookup the TLB for a specific page size
therefore looking into the appropriate set for a match

Change-Id: If77853373792d6a5ec84cf1909ee5eb567f3d0e4
Signed-off-by: Giacomo Travaglini <giacomo.travaglini@arm.com>
Reviewed-by: Andreas Sandberg <andreas.sandberg@arm.com>

giactra committed:
arch-arm: Define a SetAssociative indexing policy for the TLB

Change-Id: I8149ddc4ecf7ac3b8b7e8e1cf7eb4932fd99c34a
Signed-off-by: Giacomo Travaglini <giacomo.travaglini@arm.com>
Reviewed-by: Andreas Sandberg <andreas.sandberg@arm.com>

giactra committed:
arch-arm: Replace TLB,TLBVerbose usage in ArmMMU

Some ISAs (like Arm) have moved most of the translation logic into
the MMU and use the TLB simply as translation storage. It makes
sense to use the MMU debug flag for that logic and reduce the
scope of the TLB flag to TLB insertion/hits/misses

Change-Id: I2a164545c711d83d3e87075b0cb5c279eed274c9
Signed-off-by: Giacomo Travaglini <giacomo.travaglini@arm.com>
Reviewed-by: Andreas Sandberg <andreas.sandberg@arm.com>

giactra committed:
arch-arm: Add isValid method to the TlbEntry

Change-Id: I93b183ad0768e8afc94bb3f21387c21cdc9cc78b
Signed-off-by: Giacomo Travaglini <giacomo.travaglini@arm.com>
Reviewed-by: Andreas Sandberg <andreas.sandberg@arm.com>

giactra committed:
arch-arm: Rename TlbEntry::Lookup into TlbEntry::KeyType

KeyType definition is required if we want to store the TlbEntry
within an AssociativeCache. We could add an alias and keep the
Lookup name but this will just create extra confusion

Change-Id: Ib0b7c9529498f0f6f15ddd0e7cf3cec52966e8df
Signed-off-by: Giacomo Travaglini <giacomo.travaglini@arm.com>
Reviewed-by: Andreas Sandberg <andreas.sandberg@arm.com>

odanrc commented on src/arch/arm/ArmTLB.py:
```diff
@@ -45,12 +46,42 @@ class ArmLookupLevel(Enum):
     vals = ["L0", "L1", "L2", "L3"]
 
 
+class TlbIndexingPolicy(SimObject):
```
Consistency: TLB instead of Tlb

odanrc commented on src/arch/arm/ArmTLB.py:
```diff
@@ -45,12 +46,42 @@ class ArmLookupLevel(Enum):
+    assoc = Param.Int(Parent.assoc, "associativity")
+
+
+class TlbSetAssociative(TlbIndexingPolicy):
```
Ditto

odanrc commented on src/arch/arm/pagetable.hh:
```diff
@@ -161,31 +165,19 @@ struct V8PageTableOps64k : public PageTableOps
-struct TlbEntry : public Serializable
+struct TlbEntry;
+
+class TlbTypes
```
Ditto TLB

giactra replied:
It has to be noted that we still use TlbEntry though... What shall we do about it?

odanrc replied:
That one too. Basically, my request was to change any Tlb to TLB, to standardize the naming scheme. I understand if you would rather leave TlbEntry for the future, as it is a class that already existed before this PR, so it may generate some merge conflicts.

giactra replied:
Amended the TlbTypes. Will update TlbEntry with a separate PR

odanrc commented on src/arch/arm/pagetable.hh:
```diff
@@ -161,31 +165,19 @@ struct V8PageTableOps64k : public PageTableOps
-    {
         // virtual address
         Addr va = 0;
+        Addr pagesize = Grain4KB;
```
Should have been camel case

giactra replied:
Ack

odanrc commented on src/arch/arm/tlb.cc:
```diff
@@ -103,62 +134,19 @@ TLB::setTableWalker(TableWalker *table_walker)
+    // complete translations over partial translations
+    for (auto pagesize : observedPageSizes) {
+        lookup_data.pagesize = pagesize;
+        if (retval = table.accessEntry(lookup_data); retval)
```
IIRC, the value of an assignment is the assigned value, so is the ´; retval´ really needed?

giactra replied:
It is not but I don't like treating assignment as expressions... I prefer being explicit

odanrc commented on src/arch/arm/tlb.cc:
```diff
@@ -103,62 +134,19 @@ TLB::setTableWalker(TableWalker *table_walker)
+    // We iterate over all stored sizes, starting from the
+    // smallest until the biggest. In this way we prioritize
+    // complete translations over partial translations
+    for (auto pagesize : observedPageSizes) {
```
Should have been page_size (also, it seems that it can be const auto &)

giactra replied:
Ack

odanrc commented on src/arch/arm/tlb.hh:
```diff
@@ -167,7 +185,7 @@ class TLB : public BaseTLB
      * @return pointer to TLB entry if it exists
      */
-    TlbEntry *lookup(const Lookup &lookup_data);
+    TlbEntry *lookup(Lookup lookup_data);
```
Any particular reason to remove const &?

giactra replied:
Just because we want to make a copy of the data anyway since we iterate over different page sizes withing the lookup method and we need a non cont parameter

odanrc replied:
Oh, ok. I thought it would be something related to the fact that const references may generate bugged code (https://stackoverflow.com/a/4705871)

odanrc commented on src/arch/arm/tlb.hh:
```diff
@@ -156,9 +164,19 @@ class TLB : public BaseTLB
   public:
     using Params = ArmTLBParams;
-    using Lookup = TlbEntry::Lookup;
+    using Lookup = TlbEntry::KeyType;
```
Nit: maybe keeping the Lookup alias increases the difficulty of learning the class

giactra replied:
This PR is already big and I wanted to make it less intrusive and not so hard to review so that I didn't have to change all the Lookup instances (IIRC it is used a lot in the TLBI code). I can add an extra commit, alternatively I will open a new PR later on. Your choice

odanrc replied:
It can be left for a future PR, no worries

odanrc commented on src/arch/arm/pagetable.hh:
```diff
@@ -208,6 +200,56 @@ struct TlbEntry : public Serializable
+        Manager
+    };
+
     // Matching variables
```
Could you replace the corresponding variables by a KeyType variable?

odanrc commented on src/arch/arm/tlb.cc:
```diff
@@ -240,98 +228,75 @@ TLB::checkPromotion(TlbEntry *entry, BaseMMU::Mode mode)
 // insert a new TLB entry
 void
-TLB::insert(TlbEntry &entry)
+TLB::insert(const Lookup &lookup_data, TlbEntry &entry)
```
Why do we add another parameter instead of using entry->keyType?

giactra replied:
Interesting I will consider this option...

giactra replied:
So I looked into it and I am not fully convinced.. The issue is that KeyAccess does not convey the tag information only, but it contains access specific variables, like the aforementioned functional mentioned above. This means the TlbEntry might end up with garbage data within it

odanrc replied:
This, for me, is a hint that the KeyType may contain more stuff than it should. Nonetheless, I will disregard this request for now

giactra replied:
Thanks @odanrc !

odanrc commented on src/arch/arm/tlb.cc:
```diff
@@ -58,8 +58,38 @@ namespace gem5
+
+    for (auto candidate : candidates) {
+        auto entry = static_cast<TlbEntry*>(candidate);
+        if (entry->N == key.pagesize && entry->match(key)) {
```
Why is the entry->N == key.pagesize not part of the match check?

giactra replied:
It probably can be done but bear in mind there is more to it:

Basically at the moment the Entry::match method is not considering the page size and the reason for it is that this is quite convenient to be able to match a TLB entry not knowing the size of the page that will be matched (because I cannot tell a priori).
For example I might have a 0x8000_0000 VA and even if I know I use 4K pages, I might match a 4K page, or I could match a Block (a Huge Page in Arm, like 2MB) or I could match a partial translation (Basically an intermediate walk). Those could all be valid matches. Adding pageSize to the Lookup I am able to prioritise them; for example I could start looking for 4K pages first, and only then check for bigger sizes (a partial translation has lower priority).

There are cases where this is not needed like in TLB invalidation where I want to invalidate any potential match. So if add pageSize matching in the entry::match, I basically have to iterate over all observed page sizes when doing a TLB invalidation...

odanrc replied:
Could you add some of this info to the code? It seems relevant to help avoid bugs in the future.

giactra replied:
Done!

odanrc commented on src/arch/arm/tlb.cc:
```diff
@@ -58,8 +58,38 @@ namespace gem5
 using namespace ArmISA;
 
+TlbEntry*
+TLB::Table::accessEntry(const KeyType &key)
```
I am going to dump my thoughts, but I am unfamiliar with the code, so there may be no action point:

I think we should override the other access function instead. It would be safer, as it guarantees that no one bypasses this functional check. That said, it seems that there is no guarantee that key.functional == entry.key.functional, right?

giactra replied:
I understand your rationale. However entry.key.functional does not mean anything. The functional boolean is a property of the access, not of the entry. It means: this is not a real workload lookup/access; I just want to get the entry to be able to do a functional VA/PA translation. Therefore

1) Do not update stats
2) Do not affect replacement policy/data

Another solution which lets us avoid overriding access Entry would be to not use accessEntry for functional accesses and handle things differently in the TLB::lookup method. I am not sure it will make things better

pranith commented on src/arch/arm/ArmTLB.py:
```diff
@@ -45,6 +45,26 @@ class ArmLookupLevel(Enum):
+    cxx_template_params = ["class Types"]
+
+    # Get the size from the parent (cache)
+    size = Param.Int(Parent.size, "number of TLB entries")
```
In other contexts of indexing policies, size is defined to be MemoryType("8KB", "16KB" etc.,). Can we rename this to 'numEntries' to avoid confusion?

giactra replied:
Makes sense. I was using size for consistency with the Tlb field but I guess there's nothing wrong in doing

`num_entries = Param.Int(Parent.size, "number of TLB entries")`

pranith commented on src/arch/arm/tlb.cc:
```diff
@@ -58,8 +58,38 @@ namespace gem5
+TlbEntry*
+TLB::Table::accessEntry(const KeyType &key)
+{
+    auto entry = findEntry(key);
```
This could be:
`
if (!key.functional) {
    entry = accessEntry(key);
} else {
    entry = findEntry(key);
}
`

giactra replied:
Unless I am missing something the first case would imply an endless recursion

pranith replied:
Ah, right. Missed that we are in the same function. 😂

pranith commented on src/arch/arm/pagetable.hh:
```diff
@@ -162,7 +163,7 @@ struct V8PageTableOps64k : public PageTableOps
 
 // ITB/DTB table entry
-struct TlbEntry : public Serializable
+struct TlbEntry : public ReplaceableEntry, Serializable
```
Minor nit: can we please rename 'N' to be a bit more descriptive like 'pageSizeBits' or similar?

giactra replied:
As other comments. This makes a lot of sense but it is something historical and I would open a new PR for that, so that this one keeps being "easy" to review

giactra committed:
arch-arm: Do not include tlb.hh in mmu.hh

This commit is moving some MMU methods definition in the
source file from the header to avoid including tlb.hh

Signed-off-by: Giacomo Travaglini <giacomo.travaglini@arm.com>
Change-Id: I8fb1aeccd9c38c48b09583b4dc5d152acd09c3e6

giactra committed:
arch-arm: TranMethod is not specific to the ArmFault

It is a simple enum to distinguish between short and big
descriptors. By moving it away from the ArmFault we can
avoid including fault.hh from mmu.hh

Change-Id: Ib556b577c62f5ea3e4c8c9e0d4560a3e99c96778
Signed-off-by: Giacomo Travaglini <giacomo.travaglini@arm.com>

giactra committed:
arch-arm: Avoid unnecessary include of faults.hh

Remove unused include from self_debug.hh

Change-Id: Ic675a277ebb2ff4a319e9a7cfe2bea4af850609e
Signed-off-by: Giacomo Travaglini <giacomo.travaglini@arm.com>

giactra committed:
arch-arm: DomainType is not specific to the TlbEntry

Change-Id: I626c79973fcd60b1be36a965923999a1c9a9bc54
Signed-off-by: Giacomo Travaglini <giacomo.travaglini@arm.com>

giactra commented:
As an update, I added some refactoring commits to make unit-tests compile (sigh)

odanrc commented on src/arch/arm/pagetable.cc:
```diff
@@ -489,5 +489,14 @@ getPageTableOps(GrainSize trans_granule)
     }
 }
 
+TLBTypes::KeyType::KeyType(const TlbEntry &entry)
```
For a future PR I would recommend fixing this. The TLB entry contains a key type which we want to copy. The proper solution would be to have a KeyType var in TlbEntry and to KeyType(entry.keyType).

powerjg reviewed (approved):
lgtm. Does this affect the performance of the simulator at all?

giactra commented:
> lgtm. Does this affect the performance of the simulator at all?

Yes I am afraid.  I am currently profiling Linux boot trying to fix the performance regression (as you can see from https://github.com/gem5/gem5/pull/1739).
 
To give you an idea Linux boot with the m5exit.squashfs.arm64 disk takes 2mins and 30 seconds prior to this PR and 2mins 50 with the current PR. From flamegraph is clear the lookup logic is responsible for the regression (especially findEntry).

But more specifically I found the following motivations (there are probably more)

1) We are anyway transforming the L2 shared from being fully associative to 5-way set associative. This means we are having more conflicts in the l2shared with a partial entry potentially kicking out a valid complete translation. So we are walking more (and walking increases both guest and host time)
2) Current logic is unoptimized for fully associative TLBs (ITLB and DTLB). The indexing policy returns a set of candidates under the form of std::vector<TlbEntry*>. We are therefore copying a vector (e.g. with 64 entries) in the critical path (instruction fetch). We might need to reconsider the IP APIs to make it more FA friendly @odanrc 
3) To adhere with older design, we are copying a new TlbEntry from the table walker into the TLB while inserting a new entry. We should either use move semantics or eliminate the copy

Overall I think it is normal to introduce a small regression considered the increased complexity but I believe it shouldn't be as big as it is now...
